plugins/modules/cloud/opennebula/one_template.py

```diff
@@ -0,0 +1,265 @@
+        needs_creation = False
+        if not template and desired_state != 'absent':
+            if id:
+                module.fail_json(msg="There is no template with id=" + str(id))
+            else:
+                needs_creation = True
+
+        if desired_state == 'absent':
+            self.result = self.delete_template(template)
+        else:
+            if needs_creation:
+                if not template_data:
+                    module.fail_json(msg="To create a new template, the `template` attribute needs to be set.")
+
+                self.result = self.create_template(name, template_data)
+            else:
+                if template_data:
+                    self.result = self.update_template(template, template_data)
+                else:
+                    self.result = self.get_template_info(template)
+                    self.result['changed'] = False
+
+        self.exit()
+
+    def get_template(self, predicate):
+        # -3 means "Resources belonging to the user"
+        # the other two parameters are used for pagination, -1 for both essentially means "return all"
+        pool = self.one.templatepool.info(-3, -1, -1)
+
+        for template in pool.VMTEMPLATE:
+            if predicate(template):
+                return template
+
+        return None
+
+    def get_template_by_id(self, template_id):
+        return self.get_template(lambda template: (template.ID == template_id))
+
+    def get_template_by_name(self, template_name):
+        return self.get_template(lambda template: (template.NAME == template_name))
+
+    def get_template_instance(self, requested_id, requested_name):
+        if requested_id:
+            return self.get_template_by_id(requested_id)
+        else:
+            return self.get_template_by_name(requested_name)
+
+    def get_template_info(self, template):
+        info = {
+            'id': template.ID,
+            'name': template.NAME,
+            'template': template.TEMPLATE,
+            'user_name': template.UNAME,
+            'user_id': template.UID,
+            'group_name': template.GNAME,
+            'group_id': template.GID,
+        }
+
+        return info
+
+    def create_template(self, name, template_data):
+        if not self.module.check_mode:
+            self.one.template.allocate("NAME = \"" + name + "\"\n" + template_data)
+
+        result = self.get_template_info(self.get_template_by_name(name))
+        result['changed'] = True
+
+        return result
+
+    def update_template(self, template, template_data):
+        if not self.module.check_mode:
+            # 0 = replace the whole template
+            self.one.template.update(template.ID, template_data, 0)
+
+        result = self.get_template_info(self.get_template_by_id(template.ID))
+        # if the previous parsed template data is not equal to the updated one, this has changed
+        result['changed'] = template.TEMPLATE != result['template']
```

Comment:
So this is never `True` in `check_mode` (since the template isn't updated)?

Reply:
unfortunately yes, as I have no idea how to either parse the templated string, or to receive the raw template string via pyone (unless I overlooked something) to use it for comparison... :/

Reply:
Hmm, in that case it might be better to always return `changed=True`.

In any case, it should be documented that in check mode, change detection does not work.